Link layer keepalives

neilstephens opened the pull request:
Gday,

Here is the implementation of link layer keepalives, using link status requests. We integrated this functionality with the new link status listener, so the listener gets notified if there's a link layer timeout. After a keepalive times out, keepalives continue to be sent, and the listener is notified if the link comes back up. 

We added unit tests for all the status transitions, using confirms and not using confirms.

Now our datacon can have confirms turned off, use unsolicited and multidrop, and still fail over quickly!

Please let us know what you think.

Cheers,
Neil.

atmurray committed:
First cut at implementing link layer keep alive

atmurray committed:
Removed spurious DoSendResult from LinkStatusWait

atmurray committed:
Ensure KeepAlive timer pointer is initialised to nullptr
Reset KeepAlive timer on lower layer up.

atmurray committed:
Added some Keep Alive timer tests, not all passing

Commit:
Ensure Keep Alive PLLS states are consistent with DNP3 standard Primary Station state diagram

atmurray committed:
Fixed Keep Alive Timer pointer not being cleared after timer expiry.

Commit:
Call RetryReset prior to QueueRequestLinkStatus

Commit:
Completed unit tests for KeepAlive timer - both Reset and NonReset states

Commit:
Merge branch '2.0.x' of https://github.com/automatak/dnp3 into KeepAlive
Added call to CallStatusCallback in PLLS_LinkStatusWait final timeout

Commit:
Integrated keep alive functionality with link state listener functionality
Link state callback now indicates when link layer messages timeout
Added corresponding unit tests of link state callback to TestDNP3Manager

jadamcrain commented:
Ok, thanks. Good stuff.

I won't have any time to examine and merge until the week of June 15th due to a deadline on the 4th followed by vacation.

-AC

neilstephens committed:
Added missing state transitions in link layer primary station state machine:
- back to idle from wait states was missing for some unexpected reply function codes

neilstephens committed:
Fixed default link layer primary station OtherFunction recieved action - do nothing, not log warning.

neilstephens committed:
Link layer keep alives were putting the link layer in a sending state unbeknownst to the upper layers,
meaning the upper layers could request a send while the link layer was in the wrong state.
Issue fixed by adding a callback to the IUpperLayer interface to notify of a lower layer starting a send.

Also cleaned up some whitespace, and moved the link layer keepalive state machine code so it's a bit neater.

neilstephens commented:
During further testing, I found that the wait states (the existing reset link wait state, as well as our new link status wait) in the primary station state machine weren't transitioning back to idle if an unexpected function code was received (as the dnp3 spec says).

I added the extra transitions, and it could be back ported, because it's not specific to keep alives. It makes one unit test fail, but it's just because the unit test is a bi-directional send of reset-link, and one side gets transitioned back to an idle state via the new functionality.

I also had to add a new function to IUpperLayer to prevent higher layers from attempting to send when the link layer is in the process of doing a keep-alive.

Cheers,
Neil.

jadamcrain commented:
OK. won't have time to look until mid-month. There should be no need to tell higher layer not to send. Just have the link layer perform the send after link status completes, I.e. cache the pending   send.

neilstephens commented:
Yeah, I thought about trying to do it that way, but it seemed nicer to
utilize the existing queueing and checking in the upper layers. But
whatever fits best. Mainly I just chose a simple callback, so I was less
likely to stuff anything up.

No hurry looking at it. I wasn't expecting a reply until then anyway. I
thought you were on holiday ;-)

Cheers,
Neil.
On 06/06/2015 10:01 am, "Adam Crain" notifications@github.com wrote:

> OK. won't have time to look until mid-month. There should be no need to
> tell higher layer not to send. Just have the link layer perform the send
> after link status completes, I.e. cache the pending send.
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/automatak/dnp3/pull/127#issuecomment-109486213.
